src/compiler/checker.ts

```diff
@@ -5442,6 +5468,22 @@ namespace ts {
             target = getErasedSignature(target);
 
             let result = Ternary.True;
+            if (source.thisType && target.thisType) {
+                if (source.thisType !== voidType) {
+                    // void sources are assignable to anything.
+                    let related = compareTypes(target.thisType, source.thisType, reportErrors);
+                    if (!related) {
+                        related = compareTypes(source.thisType, target.thisType, /*reportErrors*/ false);
+                        if (!related) {
+                            if (reportErrors) {
+                                errorReporter(Diagnostics.Types_of_parameters_0_and_1_are_incompatible, "this", "this");

```

Comment:
It's always the second line of an elaboration whose first line looks like

```ts
Type '(this: D, m: number) => number' is not assignable to type '(this: C, m: number) => number'.
```

Since `this` shows up in the signatures as the first parameter, I don't think there's enough value in a custom message here.
```